packages/firestore/src/api/database.ts

```diff
@@ -893,102 +877,129 @@ export class DocumentReference<T = PublicDocumentData>
   ): Unsubscribe;
 
   onSnapshot(...args: unknown[]): Unsubscribe {
-    let options: ListenOptions = {
-      includeMetadataChanges: false
-    };
-    let currArg = 0;
-    if (
-      typeof args[currArg] === 'object' &&
-      !isPartialObserver(args[currArg])
-    ) {
-      options = args[currArg] as PublicSnapshotListenOptions;
-      currArg++;
-    }
-
-    const internalOptions = {
-      includeMetadataChanges: options.includeMetadataChanges
-    };
-
-    if (isPartialObserver(args[currArg])) {
-      const userObserver = args[currArg] as PartialObserver<
-        PublicDocumentSnapshot<T>
-      >;
-      args[currArg] = userObserver.next?.bind(userObserver);
-      args[currArg + 1] = userObserver.error?.bind(userObserver);
-      args[currArg + 2] = userObserver.complete?.bind(userObserver);
-    }
-
-    const observer: PartialObserver<ViewSnapshot> = {
-      next: snapshot => {
-        if (args[currArg]) {
-          (args[currArg] as NextFn<PublicDocumentSnapshot<T>>)(
-            this._convertToDocSnapshot(snapshot)
-          );
-        }
-      },
-      error: args[currArg + 1] as ErrorFn,
-      complete: args[currArg + 2] as CompleteFn
-    };
-
-    return firestoreClientListen(
-      this._firestoreClient,
-      newQueryForPath(this._key.path),
-      internalOptions,
-      observer
+    const options = extractSnapshotOptions(args);
+    const observer = wrapObserver<DocumentSnapshot<T>, exp.DocumentSnapshot<T>>(
```

Comment:
Above you're using import renaming to use `ExpDocumentReference` but here you're using a star import to get `exp.DocumentSnapshot`. We should do this consistently.

Reply:
Used `ExpDocumentReference`